### DISCUSSION TITLE
[Question] How do I get the value of  @punctuation.delimiter ?

### DISCUSSION THREAD

**Discussion by nabezokodaikon** [nvim-treesitter/nvim-treesitter#2938] [General]
I want to get information about the tree sitter. Among them, I want to get the value of ```@punctuation.delimiter```. How can I get it?

**Comment by nabezokodaikon** [comment]
https://github.com/nvim-treesitter/nvim-treesitter/blob/419153c1f1992ced7f9462d92f0d7505c6ff3137/queries/rust/highlights.scm#L131

In the above case, ```":":",""."``` I want to get the string ```":",";",",",""```.

**Reply by kyazdani42** [reply]
Where do you need to get the value of matches ? In a lua script ?

**Reply by theHamsta** [reply]
You want to parse the query files to do reflection on them?

**Reply by nabezokodaikon** [reply]
@kyazdani42 I intend to use denops.vim.

**Reply by nabezokodaikon** [reply]
@theHamsta This has not even been verified yet, but I am wondering if it could be used for supplemental timing for LSP.

**Reply by theHamsta** [reply]
>  supplemental timing for LSP
I don't understand how is Deno, Lsp and `;` in any way related?

You can use the `query` parser (https://github.com/nvim-treesitter/tree-sitter-query) to parse the query files themselves and then simple use `get_node_text`

**Reply by nabezokodaikon** [reply]
@theHamsta There is a plugin called ddc.vim, but I have to use regular expressions to set the LSP completion timing, so I'm wondering if I can do something about that.
Try using tree-sitter-query. Thank you.

**Reply by nabezokodaikon** [reply]
I see that tree-sitter-query is not documented. I'm going to have a hard time.

**Reply by theHamsta** [reply]
why should it be documented? It is a parser for nvim-treesitter. Just do `:TSInstall query` to install and use the playground (https://github.com/nvim-treesitter/playground) to see how the nodes are structured. It can be parse like any other language with tree-sitter.

So you want to create a completion engine that invokes the LSP at the right moments, you could also just try to complete at typical punctuation `.`?

**Reply by nabezokodaikon** [reply]
I don't really mind if you do, I just thought it would be nice to be able to do it like VSCode.
Thank you very much.

**Reply by nabezokodaikon** [reply]
It was made by the following method.

```lua
for _, match, _ in query:iter_matches(parent_node, 0, start_row, end_row) do
    for id, node in pairs(match) do
        if character_type == 'punctuation.delimiter' then
           -- Write the process here
        end
    end
end
```
